mlir/compiler-course/khovansky_d_lab4/khovansky_d_lab4.cpp

```diff
@@ -0,0 +1,77 @@
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Tools/Plugins/PassPlugin.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace mlir;
+
+namespace {
+class RemsiuiPass : public PassWrapper<RemsiuiPass, OperationPass<ModuleOp>> {
+public:
+  StringRef getArgument() const final {
+    return "RemsiuiPass_Khovansky_Dmitry_FIIT2_MLIR";
+  }
+  StringRef getDescription() const final {
+    return "Decomposes arith.remsi and arith.remui operations into equivalent "
+           "expressions using division and multiplication: a - (a // b) * b";
+  }
+
+  void runOnOperation() override {
+    ModuleOp moduleOp = getOperation();
+
+    SmallVector<Operation *, 8> remOps;
+
+    moduleOp.walk([&](Operation *op) {
+      if (isa<arith::RemSIOp>(op) || isa<arith::RemUIOp>(op)) {
+        remOps.push_back(op);
+      }
+    });
+
+    for (Operation *op : remOps) {
+      OpBuilder builder(op);
+      Location loc = op->getLoc();
+
+      // rem(a, b) = a - (a // b) * b
+      Value a, b, div, mul, result;
+
+      if (auto remsi = dyn_cast<arith::RemSIOp>(op)) {
+        a = remsi.getLhs();
+        b = remsi.getRhs();
+
+        div = builder.create<arith::DivSIOp>(loc, a, b);
+        mul = builder.create<arith::MulIOp>(loc, div, b);
+        result = builder.create<arith::SubIOp>(loc, a, mul);
+
+        remsi.replaceAllUsesWith(result);
+        remsi.erase();
+
+      } else if (auto remui = dyn_cast<arith::RemUIOp>(op)) {
+        a = remui.getLhs();
+        b = remui.getRhs();
+
+        div = builder.create<arith::DivUIOp>(loc, a, b);
+        mul = builder.create<arith::MulIOp>(loc, div, b);
+        result = builder.create<arith::SubIOp>(loc, a, mul);
+
+        remui.replaceAllUsesWith(result);
+        remui.erase();
```

Comment:
Please, move lines starting with `mul = `, `result = ` and `replaceAllUsesWith` + `erase` outside the `then` and `else` blocks to remove code duplication